File: src/main/kotlin/aris/thesis/theatricalplaysapi/specifications/PersonSpecification.java
package aris.thesis.theatricalplaysapi.specifications;

import aris.thesis.theatricalplaysapi.entities.Person;
import aris.thesis.theatricalplaysapi.specifications.base.SearchCriteria;
import org.springframework.data.jpa.domain.Specification;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;


public class PersonSpecification implements Specification<Person> {
    private SearchCriteria criteria;

    public PersonSpecification(SearchCriteria searchCriteria) {
        this.criteria = searchCriteria;
    }

    @Override
    public Predicate toPredicate(
            Root<Person> root, CriteriaQuery<?> query, CriteriaBuilder builder) {

        switch (criteria.getOperation()) {
            case EQUALITY:
                return builder.equal(root.get(criteria.getKey()), criteria.getValue());
            case NEGATION:
                return builder.notEqual(root.get(criteria.getKey()), criteria.getValue());
            case GREATER_THAN:
                return builder.greaterThan(root.<String> get(
                        criteria.getKey()), criteria.getValue().toString());
            case LESS_THAN:
                return builder.lessThan(root.<String> get(
                        criteria.getKey()), criteria.getValue().toString());
            case LIKE:
                return builder.like(root.<String> get(
                        criteria.getKey()),"%" +  criteria.getValue().toString()  + "%");
            default:
                return null;
        }
    }


}

File: src/main/kotlin/aris/thesis/theatricalplaysapi/specifications/PersonSpecification.kt
package aris.thesis.theatricalplaysapi.specifications

import aris.thesis.theatricalplaysapi.entities.Person
import aris.thesis.theatricalplaysapi.exceptions.error.never
import aris.thesis.theatricalplaysapi.exceptions.error.wrongQuery
import aris.thesis.theatricalplaysapi.specifications.base.SearchCriteria
import aris.thesis.theatricalplaysapi.specifications.base.SearchOperation
import org.springframework.data.jpa.domain.Specification
import java.lang.IllegalArgumentException
import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.CriteriaQuery
import javax.persistence.criteria.Predicate
import javax.persistence.criteria.Root

class PersonSpecification(private val criteria: SearchCriteria) : Specification<Person> {

    override fun toPredicate(root: Root<Person?>, query: CriteriaQuery<*>, builder: CriteriaBuilder): Predicate? {

        if (root.get<String>(criteria.key).equals("role") ) {

        }

        //todo remove unused operations and throw IllegalArgumentException
        return try {
            when (criteria.operation) {
                SearchOperation.EQUALITY -> builder.equal(root.get<Any>(criteria.key), criteria.value)
                SearchOperation.NEGATION -> builder.notEqual(root.get<Any>(criteria.key), criteria.value)
                SearchOperation.GREATER_THAN -> builder.greaterThan(
                    root.get(
                        criteria.key
                    ), criteria.value.toString()
                )
                SearchOperation.LESS_THAN -> builder.lessThan(
                    root.get(
                        criteria.key
                    ), criteria.value.toString()
                )
                SearchOperation.LIKE -> builder.like(
                    root.get(
                        criteria.key
                    ), "%" + criteria.value.toString() + "%"
                )
                else -> {
                    wrongQuery()
                }
            }
        } catch ( e: IllegalArgumentException) {
            throw e
        }
    }
}